src/cli/filter_arg.zig

```diff
@@ -0,0 +1,302 @@
+const std = @import("std");
+const root = @import("root");
+const bun = root.bun;
+const string = bun.string;
+const Output = bun.Output;
+const Global = bun.Global;
+const strings = bun.strings;
+const json_parser = bun.JSON;
+const Glob = @import("../glob.zig");
+
+const Package = @import("../install/lockfile.zig").Package;
+
+fn globIgnoreFn(val: []const u8) bool {
+    if (val.len == 0) {
+        return false;
+    }
+    // skip hidden directories
+    if (val[0] == '.') {
+        return true;
+    }
+    // skip node_modules
+    if (strings.eqlComptime(val, "node_modules")) {
+        return true;
+    }
+    return false;
+}
+
+const GlobWalker = Glob.GlobWalker_(globIgnoreFn, Glob.DirEntryAccessor, false);
+
+pub fn getCandidatePackagePatterns(allocator: std.mem.Allocator, log: *bun.logger.Log, out_patterns: *std.ArrayList([]u8), workdir_: []const u8, root_buf: *bun.PathBuffer) ![]const u8 {
+    bun.JSAst.Expr.Data.Store.create(bun.default_allocator);
+    bun.JSAst.Stmt.Data.Store.create(bun.default_allocator);
+
+    defer {
+        bun.JSAst.Expr.Data.Store.reset();
+        bun.JSAst.Stmt.Data.Store.reset();
+    }
+
+    var workdir = workdir_;
+
+    while (true) : (workdir = std.fs.path.dirname(workdir) orelse break) {
+        const parent_trimmed = strings.withoutTrailingSlash(workdir);
+        var name_buf: bun.PathBuffer = undefined;
+        @memcpy(name_buf[0..parent_trimmed.len], parent_trimmed);
+        name_buf[parent_trimmed.len..name_buf.len][0.."/package.json".len].* = "/package.json".*;
+        name_buf[parent_trimmed.len + "/package.json".len] = 0;
+        const json_path = name_buf[0 .. parent_trimmed.len + "/package.json".len];
+
+        log.msgs.clearRetainingCapacity();
+        log.errors = 0;
+        log.warnings = 0;
+
+        const json_file = std.fs.cwd().openFileZ(
+            name_buf[0 .. parent_trimmed.len + "/package.json".len :0].ptr,
+            .{ .mode = .read_only },
+        ) catch continue;
+        defer json_file.close();
+
+        const json_stat_size = try json_file.getEndPos();
+        const json_buf = try allocator.alloc(u8, json_stat_size + 64);
+        defer allocator.free(json_buf);
+        const json_len = try json_file.preadAll(json_buf, 0);
+        const json_source = bun.logger.Source.initPathString(json_path, json_buf[0..json_len]);
+        const json = try json_parser.ParseJSONUTF8(&json_source, log, allocator);
```

Comment:
We are parsing the package.json files multiple times here since it may happen again later 

I'm hesitant to block the PR on this but we should figure out a way to cache this to avoid the overhead 

Also, we can simplify this code a bit:

bun.sys.File.toSource(json_path, allocator);